File: 백준/Silver/18870. 좌표 압축/좌표 압축.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringBuilder sb = new StringBuilder();

        int n = Integer.parseInt(br.readLine());
        int arr[] = new int[n];
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            int num = Integer.parseInt(st.nextToken());
            arr[i] = num;
        }

        int sorted[] = Arrays.stream(arr)
                .distinct()
                .sorted()
                .toArray();

        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < sorted.length; i++) {
            map.put(sorted[i], i);
        }

        for(int i = 0; i < arr.length; i++){
            sb.append(map.get(arr[i])).append(" ");
        }

        bw.write(sb.toString());
        bw.flush();
        bw.close();
    }
}

File: 백준/Silver/18870. 좌표 압축/좌표 압축.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))
    val sb = StringBuilder()

    val n = br.readLine().toInt()
    val arr = br.readLine().split(" ").map { it.toInt() }

    val sorted = arr.distinct().sorted()

    val map = HashMap<Int, Int>()
    for(i in 0 until sorted.size) {
        map[sorted[i]] = i
    }

    for (i in 0 until arr.size) {
        sb.append(map.get(arr[i])).append(" ")
    }

    bw.write(sb.toString())
    bw.flush()
    bw.close()
}